app/src/main/java/org/thoughtcrime/securesms/video/videoconverter/MediaConverter.java

```diff
@@ -0,0 +1,355 @@
+import java.io.OutputStream;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@SuppressWarnings("WeakerAccess")
+public final class MediaConverter {
+    private static final String TAG = "media-converter";
+    private static final boolean VERBOSE = false; // lots of logging
+
+    // Describes when the annotation will be discarded
+    @Retention(RetentionPolicy.SOURCE)
+    @StringDef({VIDEO_CODEC_H264, VIDEO_CODEC_H265})
+    public @interface VideoCodec {}
+    public static final String VIDEO_CODEC_H264 = "video/avc";
+    public static final String VIDEO_CODEC_H265 = "video/hevc";
+
+    private MediaInput mInput;
+    private Output mOutput;
+
+    private long mTimeFrom;
+    private long mTimeTo;
+    private int mVideoResolution;
+    private int mVideoBitrate = 2000000; // 2Mbps
+    private @VideoCodec String mVideoCodec = VIDEO_CODEC_H264;
+    private int mAudioBitrate = 128000; // 128Kbps
+
+    private Listener mListener;
+    private boolean mCancelled;
+
+    public interface Listener {
+        boolean onProgress(int percent);
+    }
+
+    public MediaConverter() {
+    }
+
+    public void setInput(final @NonNull MediaInput videoInput) {
+        mInput = videoInput;
+    }
+
+    @SuppressWarnings("unused")
+    public void setOutput(final @NonNull File file) {
+        mOutput = new FileOutput(file);
+    }
+
+    @SuppressWarnings("unused")
+    @RequiresApi(26)
+    public void setOutput(final @NonNull FileDescriptor fileDescriptor) {
+        mOutput = new FileDescriptorOutput(fileDescriptor);
+    }
+
+    public void setOutput(final @NonNull OutputStream stream) {
+        mOutput = new StreamOutput(stream);
+    }
+
+    @SuppressWarnings("unused")
+    public void setTimeRange(long timeFrom, long timeTo) {
+        mTimeFrom = timeFrom;
+        mTimeTo = timeTo;
+
+        if (timeTo > 0 && timeFrom >= timeTo) {
+            throw new IllegalArgumentException("timeFrom:" + timeFrom + " timeTo:" + timeTo);
+        }
+    }
+
+    @SuppressWarnings("unused")
+    public void setVideoResolution(int videoResolution) {
+        mVideoResolution = videoResolution;
+    }
+
+    @SuppressWarnings("unused")
+    public void setVideoCodec(final @VideoCodec String videoCodec) throws FileNotFoundException {
+        if (selectCodec(videoCodec) == null) {
+            throw new FileNotFoundException();
+        }
+        mVideoCodec = videoCodec;
+    }
+
+    @SuppressWarnings("unused")
+    public void setVideoBitrate(final int videoBitrate) {
+        mVideoBitrate = videoBitrate;
+    }
+
+    @SuppressWarnings("unused")
+    public void setAudioBitrate(final int audioBitrate) {
+        mAudioBitrate = audioBitrate;
+    }
+
+    @SuppressWarnings("unused")
+    public void setListener(final Listener listener) {
+        mListener = listener;
+    }
+
+    @WorkerThread
+    @RequiresApi(23)
+    public void convert() throws EncodingException, IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 36. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.